pyproject.toml

```
@@ -44,6 +44,7 @@ multimethod = ">=1.8"
 python-dateutil = ">=2.8.2"
 kerchunk = { version = ">=0.1.2", optional = true }
 dask = { version = ">=2022.1.0", optional = true }
+importlib-resources = "^6.3.2"
```

Do we need to pin down to the patch version here?

---

Not necessarily, but I simply used `poetry add importlib-resources` to add it. There are other deps pinned similarly, but there is certainly no consistency.

This seems to point to a gap in the contributing guide regarding adding dependencies; not only _how_ to add them, but also _where_ to add them (other than `pyproject.toml`).

I'm happy to add a note in the contributing guide, if you can provide guidance.

Also, happy to adjust the version pin here accordingly, if there's a preference.

---

Also, a brief explanation of the _why_ (in addition to the _how_ and _where_) might be good to add to the contributing guide.

Again, happy to add all such details to the contributing guide as part of this PR, if someone can provide a few sentences.

---

There's been a bit of lively discussion around dependencies in #468, but generally, a restrictive top-pin (down to the patch version) like this will force this package to break quickly for users. So we should remove the upper bound if we don't need it.
```suggestion
importlib-resources = ">-6.3.2"
```

Similarly, is there a reason for *this* version being the lower bound?

---

> This seems to point to a gap in the contributing guide regarding adding dependencies; not only how to add them, but also where to add them (other than pyproject.toml).

Yes, agreed, we should add that. Especially since `poetry add` adds upper bounds to dependencies with `^`, which is bad for libraries (and will not change https://github.com/python-poetry/poetry/issues/3427)

---

Ah, thanks for the link to that discussion. I'll definitely dig into that to understand the thinking.

In the meantime, what can I add to the contributing guide as the basic steps for a contributor to take when adding a new dependency? What, where, and why (all very briefly)?

---

For the "what" and "where", something like this?

1. Add the dependency to `pyproject.toml` with `poetry`: `poetry add LIB` (or `poetry add --group=dev` for a development dependency)
2. On the line added to `pyproject.toml` by `poetry`, change `^` to `>=` (perhaps add a "why" bit here)
3. For a development dependency, add the same dependency and constraint to `ci/environment-dev.yaml`
4. Add the same dependency to `ci/environment-mindeps.yaml`. For a required dep, pin the exact version, but for a testing dep, leave unpinned (again, a "why" comment here might also prove helpful)

---

> There's been a bit of lively discussion around dependencies in #468, but generally, a restrictive top-pin (down to the patch version) like this will force this package to break quickly for users. So we should remove the upper bound if we don't need it.
> 
> Similarly, is there a reason for _this_ version being the lower bound?

Nope, no particular reason. That's simply what `poetry` inserted. Do we want something different?

---

> Nope, no particular reason. That's simply what poetry inserted. Do we want something different?

I think it's fine for now. Lower bounds are less problematic overall (except when they encounter incompatible upper bounds) and it's a few versions behind the latest so should be decently inclusive

---

> Those steps sound right to me. For (1) I'd be inclined to have the users just manually add it to the `pyproject.toml` instead of having to tell them to effectively manually do it again in (2), but (1) does update the lockfile too.
> 
> For 2 we could tell them to use [poetry-relax](https://github.com/zanieb/poetry-relax?tab=readme-ov-file#references) but that should probably be done in isolation with its own PR later.

My reasoning for letting poetry add the entry to pyproject.toml is because it automatically finds a version compatible with the rest of the deps, eliminating any need to manually go looking for a version number.

---

> Those steps sound right to me. For (1) I'd be inclined to have the users just manually add it to the `pyproject.toml` instead of having to tell them to effectively manually do it again in (2), but (1) does update the lockfile too.
> 
> For 2 we could tell them to use [poetry-relax](https://github.com/zanieb/poetry-relax?tab=readme-ov-file#references) but that should probably be done in isolation with its own PR later.

Cool. Can you provide the "why" in the 2 spots I mentioned? Just a sentence or 2 in each spot to provide context for contributors.

---

> What's "a few versions behind the latest"? If you're referring to importlib-resources, that version is the latest as of earlier today.

Bah, I thought I looked earlier and latest was 7.0.2; no idea what I looked at :tired_face:

---

@chuckwondo this is getting large -- perhaps we should move the contributing guide discussion to its own issue and subsequent PR?

---

As for the version, since we're still a poetry project, let's just go with:
```
importlib-resources = ">=6.3.2"
```
for now since that's what it decided and the rest were likely added that way.

---

As for the why:

> On the line added to pyproject.toml by poetry, change ^ to >= (perhaps add a "why" bit here)

because upper bounds can easily result in unsolvable user environments for Python libraries like Earthaccess. For more information, see the [poetry-relax README](https://github.com/zanieb/poetry-relax), especially the [references](https://github.com/zanieb/poetry-relax?tab=readme-ov-file#references).

> Add the same dependency to ci/environment-mindeps.yaml. For a required dep, pin the exact version, but for a testing dep, leave unpinned (again, a "why" comment here might also prove helpful)

I don't actually know the reasoning here... looks like it's because we pip install with `--no-deps` and only use the dependencies in the conda environment, so we need all dependencies reflected here. @mfisher87 @betolink got a sentence for this?